### PR TITLE
CopilotChat: Refactor deployment scripts for reusability and pipelines/workflows

### PR DESCRIPTION
**Pull request by adrianwyatt**
### Motivation and Context
- Azure storage account with CopilotChat package no longer supports anonymous public access
- Modularize deployment scripts to support devops processes, such as running locally and from pipelines/workflows.

### Description
- Moved deployment assets from copilot-chat-app/webapi/deploymenttemplates to copilot-chat-app/deploy
- Refactored to require only one bicep/ARM template (main.bicep/json)
- Separated out infrastructure and backend deployment scripts
- Added front end deployment script
- Update README to reflect changes.
- Deploy-to-Azure button will deploy WebAPI from latest CopilotChat release, aliased here:  https://aka.ms/copilotchat/webapi/latest

### CODE REVIEW

**Review by alliscode — APPROVED** [review]

**Review by hathind-ms — DISMISSED** [review]

File: samples/apps/copilot-chat-app/deploy/package-webapi.ps1 [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    # .NET framework to publish.
+    $DotNetFramework = "net6.0",
+    
+    [string]
+    # Target runtime to publish.
+    $TargetRuntime = "linux-x64",
+    
+    [string]
+    # Output directory for published assets.
+    $OutputDirectory = "$PSSCriptRoot"
```
Comment (amsacha) [inline samples/apps/copilot-chat-app/deploy/package-webapi.ps1:21]:
```suggestion
    $OutputDirectory = "$PSScriptRoot"
```

**Review by amsacha — CHANGES_REQUESTED** [review]

File: samples/apps/copilot-chat-app/deploy/main.bicep [hunk 2]
```diff
@@ -2,19 +2,15 @@
...
 Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-Bicep template for deploying Semantic Kernel to Azure as a web app service.
+Bicep template for deploying CopilotChat Azure resources.
 */
 
-@description('Name for the deployment - Must consist of alphanumeric characters or \'-\'')
-param name string = 'semkernel'
+@description('Name for the deployment consisting of alphanumeric characters or dashes (\'-\')')
+param name string = 'copichat'
```
[the remark this answers is not in the document]
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/deploy/main.bicep:9]: We need to keep the character count down for some of the names, especially CosmosDB account names.

File: samples/apps/copilot-chat-app/deploy/deploy-azure.ps1 [hunk 3]
```diff
@@ -73,17 +57,39 @@ param(
...
-    `\`"packageUri`\`": { `\`"value`\`": `\`"$PackageUri`\`" },
-    `\`"appServiceSku`\`": { `\`"value`\`": `\`"$AppServiceSku`\`" },
-    `\`"semanticKernelApiKey`\`": { `\`"value`\`": `\`"$SemanticKernelApiKey`\`" },
+    `\`"webAppServiceSku`\`": { `\`"value`\`": `\`"$WebAppServiceSku`\`" },
+    `\`"aiService`\`": { `\`"value`\`": `\`"$AIService`\`" },
+    `\`"aiApiKey`\`": { `\`"value`\`": `\`"$AIApiKey`\`" },
+    `\`"aiEndpoint`\`": { `\`"value`\`": `\`"$AIEndpoint`\`" },
+    `\`"deployNewAzureOpenAI`\`": { `\`"value`\`": $(If ($DeployAzureOpenAI) {"true"} Else {"false"}) },
     `\`"deployQdrant`\`": { `\`"value`\`": $(If (!($NoQdrant)) {"true"} Else {"false"}) },
     `\`"deployCosmosDB`\`": { `\`"value`\`": $(If (!($NoSpeechServices)) {"true"} Else {"false"}) },
```
[the remark this answers is not in the document]
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/deploy/deploy-azure.ps1:94]: nice catch! The double-negative are to support default values for switches. Happy to discuss more in for a future PR.

File: samples/apps/copilot-chat-app/deploy/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+chmod +x ./package-webapi.sh
+./package-webapi.sh
+
+chmod +x ./deploy-webapi.sh
+./deploy-webapi.sh --subscription {YOUR_SUBSCRIPTION_ID} --resource-group rg-{YOUR_DEPLOYMENT_NAME} --deployment-name {YOUR_DEPLOYMENT_NAME}
+```
+
+
+# Deploy Frontend (WebApp)
+**TODO application ID**
```
Comment (gitri-ms) [inline samples/apps/copilot-chat-app/deploy/README.md:74]: Update this before checking in?

File: .github/workflows/dotnet-copilotchat-ci.yml [hunk 5]
```diff
@@ -0,0 +1,76 @@
+#
+# This workflow will package the Copilot Chat application for deployment.
+#
+
+name: dotnet-pr
+
+on:
+  pull_request:
+    branches: [ "main", "feature*", "sk-piplines" ]
```
Comment (gitri-ms) [inline .github/workflows/dotnet-copilotchat-ci.yml]: typo: "sk-pipelines"

File: samples/apps/copilot-chat-app/deploy/deploy-azure.sh [hunk 6]
```diff
@@ -1,25 +1,21 @@
 #!/bin/bash
 
-# Creates a Semantic Kernel service deployment using an existing Azure OpenAI account.
+# Deploy CopilotChat Azure resources.
 
 set -e
 
 usage() {
-    echo "Usage: $0 -d DEPLOYMENT_NAME -s SUBSCRIPTION -e ENDPOINT -o AZURE_OPENAI_API_KEY [OPTIONS]"
+    echo "Usage: $0 -d DEPLOYMENT_NAME -s SUBSCRIPTION --ai AI_SERVICE_TYPE -aikey AI_SERVICE_KEY [OPTIONS]"
```
Comment (gitri-ms) [inline samples/apps/copilot-chat-app/deploy/deploy-azure.sh:8]: Should just be -ai, not --ai

File: samples/apps/copilot-chat-app/deploy/deploy-webapi.ps1 [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    $ResourceGroupName,
+    
+    [Parameter(Mandatory)]
+    [string]
+    # Name of the previously deployed Azure deployment 
+    $DeploymentName,
+
+    [string]
+    # CopilotChat WebApi package to deploy
+    $PackageFilePath = "$PSSCriptRoot/out/webapi.zip"
```
Comment (gitri-ms) [inline samples/apps/copilot-chat-app/deploy/deploy-webapi.ps1:24]: nit: casing: PSScriptRoot

File: samples/apps/copilot-chat-app/deploy/deploy-webapp.ps1 [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+$webapiApiKey=($(az webapp config appsettings list --name $webapiName --resource-group $ResourceGroupName | ConvertFrom-JSON) | Where-Object -Property name -EQ -Value Authorization:ApiKey).value
+Write-Host "webappUrl: $webappUrl"
+Write-Host "webappName: $webappName"
+Write-Host "webapiName: $webapiName"
+Write-Host "webapiUrl: $webapiUrl"
+
+# Set UTF8 as default encoding for Out-File
+$PSDefaultParameterValues['Out-File:Encoding'] = 'ascii'
+
+$envFilePath="$PSSCriptRoot/../webapp/.env"
```
Comment (gitri-ms) [inline samples/apps/copilot-chat-app/deploy/deploy-webapp.ps1:56]: nit: casing: PSScriptRoot (also lines 63, 65, 66, 68)

File: samples/apps/copilot-chat-app/deploy/package-webapi.ps1 [hunk 9]
```diff
@@ -0,0 +1,48 @@
+<#
+.SYNOPSIS
+Package CiopilotChat's WebAPI for deployment to Azure
```
Comment (gitri-ms) [inline samples/apps/copilot-chat-app/deploy/package-webapi.ps1:3]: typo: CiopilotChat => CopilotChat

**Review by gitri-ms — CHANGES_REQUESTED** [review]

**Review by amsacha — APPROVED** [review]

**Review by amsacha — DISMISSED** [review]

**Review by hathind-ms — APPROVED** [review]

**Pull request merged**